### DISCUSSION TITLE
Visible marking of ABC implementation possibility

### DISCUSSION THREAD

**Discussion by MRiabov** [microsoft/pylance-release#6528] [Enhancement]
Hello,
Related to, automatic inherited method implementation of `ABC` and `ABCMeta, I would suggest a visible marking a possibility of such action, e.g. a warning or a spelling underline under the class. It is only natural that abstract class needs an implementation. E.g:

`class SketchEnv(Abstract

Right now there is no marking that this is possible, and new users (me) don't know that this is a thing.

Thank you.

_P.S._ Also, it would be good if an ABCMeta implementation of ABC would have their methods declated as @abstractmethod automatically, e.g.

```
 class AbstractClass(ABC):
        @abstractmethod        
        def desiredChildMethod():
            pass
            
    class MetaAbstract(ABCMeta, AbstractClass):

        @abstractmethod     _# @abstractmethod inserted automatically_
        def desiredChildMethod():
              _#this passes the method forward_
            pass
```
I know this isn't always the case (with real implementations), but my guess is that it is more often that not.

P.S.S. Better context (21.6.2024): https://github.com/microsoft/pylance-release/issues/5362

**Comment by heejaechang** [comment]
tagging @StellaHuang95

**Comment by rchiodo** [comment]
Adding the GIF that shows how to trigger this:

![image](https://github.com/microsoft/pylance-release/assets/19672699/67811da3-4610-4020-827e-8ebdd13fda35)

I believe this new request is to mark the class somehow so that the code action is more apparent?

**Comment by MRiabov** [comment]
> Adding the GIF that shows how to trigger this:
> 
> ![image](https://private-user-images.githubusercontent.com/19672699/342828134-67811da3-4610-4020-827e-8ebdd13fda35.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMzkzMjIsIm5iZiI6MTcxOTMzOTAyMiwicGF0aCI6Ii8xOTY3MjY5OS8zNDI4MjgxMzQtNjc4MTFkYTMtNDYxMC00MDIwLTgyN2UtOGViZGQxM2ZkYTM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDE4MTAyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3ZDkxZjQwMGQ2OGNmMjMyMTA0N2JjNmI1ZmNiNjFiZTI3NTY5MGY1ZDliYzdkYmYwOTFkYzRjZDIxYmIzZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.D9_A8NzRcvv4stwZHCwm63-GGEM0PVSroQZ1T3lEF1k)
> 
> I believe this new request is to mark the class somehow so that the code action is more apparent?

Precisely. It's simply non-obvious that something can be done.
Btw, the attachment you have is image, not GIF.

**Comment by MRiabov** [comment]
bump?

**Comment by rchiodo** [comment]
It's unlikely we'd get to this anytime soon. It would need a lot of people asking for it.

**Comment by rchiodo** [comment]
And it's not obvious how we should expose it. I'm not sure a squiggle makes sense. Those indicate something wrong with your code.

This is more likely a VS code UI suggestion. How to indicate Code Actions are available on different pieces of code without clicking on them.

**Comment by MRiabov** [comment]
Well, squiggle can be white, as in grammar typo underline.
And indeed, something _is_ wrong - as we are not implementing the necessary for the class functions - although we should.

**Comment by erictraut** [comment]
If you are not implementing all of the abstract methods, pyright (the type checker upon which pylance is built) will emit an error if you attempt to instantiate the incomplete class. It will tell you which methods are unimplemented. Furthermore, if you mark the concrete class as `@final` (using the `typing.final` decorator), pyright will tell you about the missing implementations without an attempt to instantiate the class. To get these diagnostics, you need to enable the `reportAbstractUsage` diagnostic check.

```python
from abc import ABC, abstractmethod
from typing import final

class AbstractClass(ABC):
    @abstractmethod
    def method1(self) -> int:
        raise NotImplementedError()

class ConcreteClass1(AbstractClass):
    pass

c = ConcreteClass1() # Error: cannot instantiate abstract class

@final
class ConcreteClass2(AbstractClass): # Error: final class does not implement abstract symbols
    pass
```

![image](https://github.com/microsoft/pylance-release/assets/7040122/8e091924-22e3-4ffd-bab2-6578774cafed)
